### PR TITLE
!str #18737: Migration guide from 1.0

### PR DESCRIPTION
**Pull request by drewhk**
Preliminar version of the migration guide

### CODE REVIEW

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+Update procedure
+----------------
+
+1. Replace all uses of ``Flow.wrap`` when it converts a ``Graph`` to a ``Flow`` with ``Flow.fromGraph``
+2. Replace all uses of ``Flow.wrap`` when it converts a ``Source`` and ``Sink`` to a ``Flow`` with
+   ``Flow.fromSinkAndSource`` or ``Flow.fromSinkAndSourceMat``
+3. Replace all uses of ``BidiFlow.wrap`` when it converts a ``Graph`` to a ``BidiFlow`` with ``BidiFlow.fromGraph``
+4. Replace all uses of ``BidiFlow.wrap`` when it converts two ``Flow``s to a ``BidiFlow`` with
+   ``BidiFlow.fromFlows`` or ``BidiFlow.fromFlowsMat``
+5. Repplace all uses of ``BidiFlow.apply()`` (Scala DSL) or ``BidiFlow.create()`` (Java DSL) when it converts two
```
Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: Replace

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+FlowGraph builder methods have been renamed
+===========================================
+
+There is now only one graph creation method called ``create`` which is analogous to the old ``partial`` method. For
+closed graphs now it is explicitly required to return ``ClosedShape`` at the end of the builder block.
+
+Update procedure
+----------------
+
+1. Replace all occurrences of ``FlowGraph.create()`` with ``FlowGraph.partial()``
```
Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: Replace all occurrences of ``FlowGraph.partial()`` with ``FlowGraph.create()``

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+2. Add ``ClosedShape`` as a return value of the builder block
```
Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: Replace all occurrences of ``FlowGraph.closed()`` with ``FlowGraph.create()`` and add ``ClosedShape`` as a return value of the builder block

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+1. Replace all uses of ``Flow.empty()`` with ``Flow.create``.
+
+TODO: code example
+
+``flatten(FlattenStrategy)`` has been replaced by named counterparts
+====================================================================
+
+To simplify type inference in Java 8 and to make the method more discoverable, ``flatten(FlattenStrategy.concat)``
+has been removed and replaced with the alternative method ``flatten(FlattenStrategy.concat)``.
```
Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: `flattenConcat()`

Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: Btw, that PR doesn't look like it's been merged yet.

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+TODO: code example
+
+FlexiMerge an FlexiRoute has been replaced by GraphStage
+========================================================
+
+The ``FlexiMerge`` and ``FlexiRoute`` DSLs have been removed since they provided an abstraction that was too limiting
+and a better abstraction have been created which is called ``GraphStage``. ``GraphStage`` can express fan-in and
+fan-out stages, but many other constructs as well with possibly multiple input and output ports (e.g. a ``BidiStage``).
+
+This new abstraction provides a more uniform way to crate custom stream processing stages of arbitrary ``Shape``. In
```
Comment (viktorklang) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: crate => create

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst (hunk 4 above)
Comment (drewhk) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: this is against my branch so far, and just for early preview. I had no time to fully proof-read it and create code samples.

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+read the* ``GraphStage`` *documentation (TODO) for details.*
+
+Variance of Inlet and Outlet (Scala DSL)
+========================================
+
+Scala uses *declaration site variance* which was cumbersome in the cases of ``Inlet`` and ``Outlet`` as they are
+purely symbolic object containing no fields or methods and which are used both in input and output locations (wiring
+an ``Outlet`` into an ``Inlet``; reading in a stage from an ``Inlet``). Because of this reasons all users of these
+port abstractions now use *use-site variance* (just like Java variance works). This in general does not affect user
+code expect the case of custom shapes, which now require ``@uncheckedVariance`` annotations on their ``Inlet`` and
```
Comment (rkuhn) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: `s/expect/except/`

File: akka-docs-dev/rst/migration-guide-1.0-2.x.rst [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+The ``isHoldingUpstream`` method used to return true if the upstream port was in holding state and a completion arrived
+(inside the ``onUpstreamFinished`` callback). Now it returns ``false`` when the upstream is completed.
+
+Update procedure
+----------------
+
+1. Those stages that relied on the previous behavior need to introduce an extra ``Boolean`` field with initial value
+   ``false``
+2. This field must be set on every call to ``holdUpstream()`` (and variants).
+3. In completion, instead of calling ``isHoldingUpstream`` read this variable instead.
```
Comment (rkuhn) [inline akka-docs-dev/rst/migration-guide-1.0-2.x.rst]: mention another possibility: rewrite as GraphStage

**Comment by rkuhn** [comment]
good start, thanks!

**Pull request closed without merging**